source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+--------
+
+The `SessionOptions <../sessions/driver-sessions.rst#mongoclient-changes>`_
+used to construct explicit `ClientSession
+<../sessions/driver-sessions.rst#clientsession>`_ instances MUST accept a new
+``defaultTimeoutMS`` option, which specifies the ``timeoutMS`` value for the
+following operations executed on the session:
+
+#. commitTransaction
+#. abortTransaction
+#. withTransaction
+#. endSession
+
+If this option is not specified for a ``ClientSession``, it MUST inherit the
+``timeoutMS`` of its parent MongoClient.
+
+Convenient Transactions API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If ``timeoutMS`` is set, drivers MUST apply it to the entire
+``withTransaction`` call. To propagate the timeout to the user-supplied
+callback, drivers MUST store the timeout as a field on the ClientSession
+object. This field SHOULD be private to ensure that a user can not modify it
+while a ``withTransaction`` call is in progress. Drivers that cannot make
+this field private MUST signal that the field should not be accessed or
+modified by users if there is an idiomatic way to do so in the language (e.g.
+underscore-prefixed variable names in Python) and MUST document that
+modification of the field can cause unintended correctness issues for
+applications. Drivers MUST document that the remaining timeout will not be
+applied to callback operations that do not use the ClientSession and that
+overridding ``timeoutMS`` for operations inside the provided callback will
+result in a client-side error. If the callback returns an error and the
+transaction must be aborted, drivers MUST refresh the ``timeoutMS`` value for
+the ``abortTransaction`` operation.
+
+If ``timeoutMS`` is not set, drivers MUST continue to exhibit the existing
+120 second timeout behavior. Drivers MUST NOT change existing implementations
+to use ``timeoutMS=120000`` for this case.
+
+See `withTransaction communicates timeoutMS via ClientSession`_ and
+`withTransaction refreshes the timeout for abortTransaction`_.
+
+GridFS API
+----------
+
+GridFS buckets MUST inherit ``timeoutMS`` from their parent MongoDatabase
+instance and all methods in the GridFS Bucket API MUST support the
+``timeoutMS`` option. For methods that create streams (e.g.
+``open_upload_stream``), the option MUST cap the lifetime of the entire
+stream. This MUST include the time taken by any operations executed during
+stream construction, reads/writes, and close/abort calls. For example, if a
+stream is created at time ``T``, the final ``close`` call on the stream MUST
+finish all blocking work before time ``T + timeoutMS``. Methods that interact
+with a user-provided stream (e.g. ``upload_from_stream``) MUST use
+``timeoutMS`` as the timeout for the entire upload/download operation. If the
+user-provided streams do not support timeouts, drivers MUST document that the
+timeout for these methods may be breached if calls to interact with the
+stream take longer than the remaining timeout. If ``timeoutMS`` is set, all
```

Comment:
The `timeoutMS` option should always apply to the entire operation. However, we recognized that this is likely not possible for many drivers that use standard library stream types/functions as PHP does. In the case you described, PHP would have to document that `timeoutMS` may elapse when interacting with the user-provided stream, but this wouldn't be detected until the stream interaction is complete. I wouldn't say that it has no bearing, though, because the time taken by these interactions will still count against `timeoutMS`; we just won't detect it right away. For example, the following could happen:

1. Call `uploadFromStream` with `timeoutMS=1000`
2. Driver tries to read from user-provided stream. The user's stream takes 2 seconds to return the data.
3. Driver executes an `insert` command to upload chunks into the database. This fails quickly because the upload's timeout has expired.

This goes back to what I mentioned in some other threads about storing the timeout as a deadline: in this scenario, the deadline expires, but isn't checked until the driver has regained control from the user-provided stream type.